man/intronProspector.1
.\" Automatically generated by Pandoc 3.8.3
.\"
.TH "intronProspector" "1" "April 02, 2026" "Call intron junctions"
.SH NAME
\f[B]intronProspector\f[R] \(em Extract putative intron junctions from
RNA\-Seq alignments
.SH SYNOPSIS
\f[CR]intronProspector [options] [readaligns ...]\f[R]
.SH DESCRIPTION
Find putative intron junctions in a RNA\-Seq alignment.
The \f[I]readaligns\f[R] files maybe in SAM, BAM, or CRAM format and
does not need to be sorted or indexed and maybe streamed.
If omitted, stdin is used.
.PP
This program allows for integrating splice junction calling into an
alignment pipeline.
Pass\-through mode copys the alignment file on \f[CR]stdin\f[R] to
\f[CR]stdout\f[R].
It can then sit between an aligner outputting SAM and
\f[CR]samtools\f[R] that converts to BAM/CRAM.
.PP
Both short\-read (Illumina) and long\-read RNA\-Seq can be process.
For short\-reads, it is recommended to use
\f[CR]\-\-min\-confidence\-score=1.0\f[R], which is the default.
For long\-read experiments this should be disable
with\f[CR]\-\-min\-confidence\-score=0.0\f[R].
Low coverage experiment may wish to disable or use a lower confidence
filter.
Introns are as determined by the aligner and indicated in the BAM by
\f[CR]N\f[R] operations.
.PP
If a genome is provided, only canonical introns are kept (AT/AG, GC/AG,
and AT/AC).
This can be overridden with \f[CR]\-\-sj\-filter=all\f[R]
.PP
Process multiple BAMs at once will use more more memory, as intron
counts are not collected per chromosome.
This mode does not require a sorted BAM and is incompatible with
\f[CR]\-\-pass\-through\f[R].
It may also improve the confidence scoring in some cases.
.PP
The \f[CR]intronProspectorMerge\f[R] program can be used to convert to
from the \f[CR]\-\-intron\-calls\f[R] format to other formats as well as
merge the output from multiple \f[CR]intronProspector\f[R] runs.
.PP
TSV and BED files will be automatically compressed with \f[CR]gzip\f[R]
if they end in \f[CR].gz\f[R].
.SH OPTIONS
\f[CR]\-h, \-\-help\f[R]
.RS
.PP
Prints brief usage information.
.RE
.PP
\f[CR]\-v, \-\-version\f[R]
.RS
.PP
Prints the current version number.
.RE
.PP
\f[CR]\-g fasta, \-\-genome\-fasta=fasta\f[R]
.RS
.PP
Genome FASTA file, must be indexed by \f[CR]samtools faidx\f[R].
Donor and acceptor dinuclotides are identified if provided.
THIS SHOULD NORMALLY BE PROVIDE!
.RE
.PP
\f[CR]\-S, \-\-skip\-missing\-targets\f[R]
.RS
.PP
If a target sequence is not in the genome, skip the alignment rather
than generate an error.
.RE
.PP
\f[CR]\-f spec, \-\-sj\-filter=spec\f[R]
.RS
.PP
Filter based on splice junctions motif.
A spec of \f[CR]canon\f[R] keeps only canonical splice junctions (GT/AG,
GC/AG, and AT/AC).
A spec of \f[CR]all\f[R] passes all through.
If \f[CR]canon\f[R] is specified, a genome must be supplied, and if a
genome is supplied, \f[CR]canon\f[R] becomes the default.
.RE
.PP
\f[CR]\-a INT, \-\-min\-anchor\-length=INT\f[R]
.RS
.PP
Minimum anchor length.
Junctions which satisfy a minimum anchor length on both sides are
reported.
Mismatch bases don\(cqt count towards the meeting this threshold.
The default is 8 bases.
.RE
.PP
\f[CR]\-i INT, \-\-min\-intron\-length=INT\f[R]
.RS
.PP
Minimum intron length.
The default is 70 bases.
.RE
.PP
\f[CR]\-I INT,  \-\-max\-intron\-length=INT\f[R]
.RS
.PP
Maximum intron length.
The default is 500000 bases.
.RE
.PP
\f[CR]\-d,  \-\-allow\-anchor\-indels\f[R]
.RS
.PP
Are indels allowed in anchors?
Useful for ONT reads.
Indels bases don\(cqt count towards the meeting the anchor size
threshold.
The default is don\(cqt allow indels.
\&.
.RE
.PP
\f[CR]\-m INT,  \-\-max\-anchor\-indel\-size=INT\f[R]
.RS
.PP
Maximum size of any contiguous indel in an anchor if indels are allowed
in anchors.
Exceeding this length discards the intron.
The default is 36.
.RE
.PP
\f[CR]\-C FLOAT, \-\-min\-confidence\-score=FLOAT\f[R]
.RS
.PP
Calculate the Shannon\-Wiener Diversity Index to use as a confidence
score and discard intron calls below this value.
A value of 1.0 is a good threshold for filtering.
A value of 0.0 disables filtering.
The default is 1.0.
This methodology it taken from JuncBASE DOI: 10.1101/gr.108662.110.
.RE
.PP
\f[CR]\-r INT, \-\-min\-read\-count=INT\f[R]
.RS
.PP
Minimum number of reads supporting a junction for it to be reported.
Junctions with fewer reads than this threshold are discarded.
The default is 0 (no filtering).
.RE
.PP
\f[CR]\-s STRING, \-\-strandness=STRING\f[R]
.PP
\f[CR]\-u, \-\-unsorted\f[R]
.RS
.PP
SAM/BAM files are not sorted.
This will require more memory, as intron counts are not collected per
chromosome.
.RE
.RS
.PP
Strand specificity of RNA library preparation.
Use \f[CR]UN\f[R] for unstranded, \f[CR]RF\f[R] for first\-strand,
\f[CR]FR\f[R] for second\-strand (case\-insensitive).
The default is \f[CR]UN\f[R].
This is used to set the strand in the junction\-format BED file.
.RE
.PP
\f[CR]\-X category, \-\-exclude=category\f[R]
.RS
.PP
Exclude reads or introns from this category.
Current categories are: \f[I]multi\f[R] \- excluding multi\-mapped
reads.
.RE
.PP
\f[CR]\-\-set\-XS\-strand\-tag\f[R]
.RS
.PP
Set the XS:A tag to the strand if the strand for a read can be
determined from the introns based on a count of recognized vs
unrecognized splice sites.
If the strand can\(cqt be determined, the tags are not modified,
possible leaving an existing XS:A tag in place.
Requires \f[CR]\-\-pass\-through\f[R] and \f[CR]\-\-genome\-fasta\f[R].
.RE
.PP
\f[CR]\-\-set\-TS\-strand\-tag\f[R]
.RS
.PP
Set the TS:A tag in the same manner as
\f[CR]\-\-set\-XS\-strand\-tag\f[R].
.RE
.RS
.PP
Skip getting splice junctions when target sequence is missing in genome
FASTA rather than generate an error and stop.
.RE
.PP
\f[CR]\-c FILE, \-\-intron\-calls=FILE\f[R]
.RS
.PP
Write a tab\-separated values (TSV) file, with the junctions calls.
It will contain the following columns : * chrom \- chromosome name *
intron_start \- zero based starting coordinates of intron.
* intron_end \- zero based ending coordinates of intron.
* strand \- \f[CR]+\f[R], \f[CR]\-\f[R], or \f[CR].\f[R] if not known.
This is set only if experiment is stranded, it is not set from genomic
sequence.
* uniq_mapped_count \- number of uniquely mapped reads supporting the
junction.
* multi_mapped_count \- number of uniquely mapped reads supporting the
junction.
* unsure_mapped_count \- number of reads that are either in discordant
or partial mapped pair or do not have the BAM \f[CR]NH\f[R] tag.
* max_left_overhang \- maximum number of bases overlapping the exon
upstream of the intron.
* max_right_overhang \- maximum number of bases overlapping the exon
downstream of the intron.
* confidence \- confidence score.
* splice_sites \- Donor\-acceptor pair, if genome provided.
In upper\-case for known splicing patterns, lower\-case for unknown.
If strand is not known, this field is empty.
.RE
.PP
\f[CR]\-j FILE, \-\-junction\-bed=FILE\f[R]
.RS
.PP
Write the junction calls and support anchors to this BED 12 file.
This is in the same format as TopHat \f[CR]junctions.bed\f[R] and
\f[CR]regtools junction extract\f[R] output.
It is a UCSC BED track, with each junction consists of two connected BED
blocks, where each block is as long as the maximal overhang of any read
spanning the junction.
The score is the number of alignments spanning the junction, with a
maximum score of 1000 for UCSC browser compatibility.
If genome is supplied, BED is colored green for U2 junctions, blue for
U12, or red for unknown.
.RE
.PP
\f[CR]\-n FILE, \-\-intron\-bed=FILE\f[R]
.RS
.PP
Write the intron BED 9 with the bounds of the introns.
The score is the number of alignments spanning the junction, with a
maximum score of 1000 for UCSC browser compatibility.
If genome is supplied, BED is colored green for U2 junctions, blue for
U12, or red for unknown.
.RE
.PP
\f[CR]\-b FILE, \-\-intron\-bed6=FILE\f[R]
.RS
.PP
Write the intron BED 6 with the bounds of the introns.
The score is the number of alignments spanning the junction.
This is for software not wanting to create a browser track.
.RE
.PP
\f[CR]\-p FILE, \-\-pass\-through=FILE\f[R]
.RS
.PP
Pass through input BAM/SAM records to this file, used for constructing
pipelines with \f[CR]/dev/stdout\f[R] is specified.
CRAM output is not support, these will be written as BAM.
.RE
.PP
\f[CR]\-D FILE, \-\-debug\-trace=FILE\f[R]
.RS
.PP
Output records, in TSV format, for reach read intron indicating the
information going into classifying it, including read name.
First few columns are BED\-like for easy conversion.
.RE
.SH NOTES
The computation of strand is problematic.
If the strandness of the experiment is specified, then that is used to
determine stand.
If the alignment provides the XS attribute, that is used.
Otherwise, the strand can\(cqt be determined from the BAM.
If the genome is provided and a known splice sites are detected, this is
then used if the stand is not identified by other methods.
.PP
Secondary alignments are not used to support introns.
.SH EXAMPLES
Call junctions from a BAM file, also creating BEDs of junctions and
introns:
.IP
.EX
intronProspector \-\-genome\-fasta=thegenome.fa.gz \-\-intron\-calls=introns.tsv \-\-junction\-bed=juncs.bed \-\-intron\-bed=introns.bed reads.bam
.EE
.PP
Pipeline to call introns and create a CRAM file:
.IP
.EX
cat reads.sam \(rs
    | samtools sort \-O sam  \(rs
    | ./intronProspector \-c introns.tsv \-\-genome\-fasta=thegenome.fa.gz \-p /dev/stdout \(rs
    | samtools view \-O CRAM \-T grch38.fa >reads.cram
.EE
.PP
Note that the \f[CR]cat\f[R] command could be an aligner outputting a
SAM file and that the genome FASTA file must be index by
\f[CR]samtools faidx\f[R].
.SH BUGS
See GitHub Issues: \c
.UR https://github.com/diekhans/intronProspector/issues
.UE \c
.SH AUTHOR
Mark Diekhans \c
.MT markd@ucsc.edu
.ME \c
.PP
Source available from \c
.UR https://github.com/diekhans/intronProspector
.UE \c
.PP
Base on code from RegTools \c
.UR https://github.com/griffithlab/regtools
.UE \c
\ by Avinash Ramu \c
.MT aramu@genome.wustl.edu
.ME \c
\&.
.SH AUTHORS
Mark Diekhans.


CHANGELOG.md
# Major and user-visible changes

## version 1.5.0 2026-04-04
* Add --min-read-count option to filter introns based on number of supporting reads.
* If htslib is not found by configure, error message makes suggestion when using Conda.
* The intronProspector and intronProspectorMerge commands have been renamed to 
  intron-prospector and intron-prospector-merge to follow more common conventions and
  decrease wear on the shift key.  Wrapper scripts are provide for compatibility, which
  will be removed in the next major release.

## version 1.4.0 2025-08-12
* Add --sj-filter=canon option `intronProspector` to keep only canonical splice junctions.
  This is the default if a genome is provide.  Filtering can be disabled with
  --sj-filter=all. [incompatibility]
* Add --sj-filter=canon option `intronProspectorMerge` to keep only canonical splice
  junctions.
* Make --min-confidence-score=1.0 the default. May want to lower or disable for low-coverage
  or long-reads. [incompatibility]


## version 1.3.0 2025-04-10
* Added ability to read and write gzip compressed TSV and BED files based on an name
  extension of '.gz'


## version 1.2.0 2025-03-15
* Added ability to process multiple BAMs at once.  This will use more memory
  and possibly improve the confidence scoring in some cases
* Added option to allow input BAM to be unsorted.  This will use more memory.


src/intronProspector.cc
/* intronProspector.cc - command line program to extract 

    Copyright (c) 2018-2025, Mark Diekhans, University of California, Santa Cruz

This code is derived from the regtools package available at:

  https://github.com/griffithlab/regtools

We thank them for providing this software under a flexible license:

    Copyright (c) 2015, The Griffith Lab

    Author: Avinash Ramu <aramu@genome.wustl.edu>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.  */

#include <getopt.h>
#include <algorithm>
#include <vector>
#include "AutoGzip.hh"
#include "junctions_extractor.hh"
#include "strconvert.hh"
#include "genome.hh"
#include "version.hh"

using namespace std;

static const uint32_t DEFAULT_MIN_ANCHOR_LENGTH = 8;
static const uint32_t DEFAULT_MIN_INTRON_LENGTH = 70;
static const uint32_t DEFAULT_MAX_INTRON_LENGTH = 500000;
static const bool DEFAULT_ALLOW_ANCHOR_INDELS = false;
static const uint32_t DEFAULT_MAX_ANCHOR_INDEL_SIZE = 36;
static const float DEFAULT_MIN_CONFIDENCE_SCORE = 1.0;
static const uint32_t DEFAULT_MIN_READ_COUNT = 0;
static const Strandness DEFAULT_STRANDED = UNSTRANDED;

static const char *usage_msg =
#include "intronProspector.man.h"
    ;

// Usage statement for this tool
static void usage() {
    cerr << usage_msg;
    exit(0);
}

// error and reference to -h
static void cmd_error(const string& msg) {
    cerr << "Error: "  << msg << endl;
    cerr << "Use --help to get usage" << endl;
    exit(1);
}


// open a output file if not empty, otherwise return NULL
static inline ostream* open_out_or_null(const string& fname) {
    return (fname != "") ? new AutoGzipOutput(fname) : NULL;
}

// convert a specification for strandness to constant.
static Strandness str_to_strandness(const string s) {
    // case-insensitive compare
    string su(s);
    transform(su.begin(), su.end(), su.begin(), ::toupper);
    if (su == "UN") {
        return UNSTRANDED;
    } else if (su == "RF") {
        return RF_STRANDED;
    } else if (su == "FR") {
        return FR_STRANDED;

    } else {
        cerr << "Error: invalid strandness value '" + s + "', expected one of 'UN', 'RF', or 'FR' (case insensitive)" << endl;
        exit(1);
    }
}

// parse an exclude category
static unsigned parse_exclude_cat(const string &cat) {
    if (cat == "multi") {
        return EXCLUDE_MULTI;
    } else {
        cerr << "Error: invalid exclude category '" << cat << "', expected one of: 'multi'" << endl;
        exit(1);
    }
}

// Parse command line
class CmdParser {
    public:
    // input
    vector<string> bam_files;
    uint32_t min_anchor_length;
    uint32_t min_intron_length;
    uint32_t max_intron_length;
    bool allow_anchor_indels;
    bool unsorted;
    uint32_t max_anchor_indel_size;
    float min_confidence_score;
    uint32_t min_read_count;
    JunctionFilter junction_filter;
    Strandness strandness;
    unsigned excludes;
    string genome_fa;
    bool skip_missing_targets;

    // output
    string junction_bed;
    string intron_bed;
    string intron_bed6;
    string intron_call_tsv;
    string bam_pass_through;
    string debug_trace_tsv;
    bool set_XS_strand_tag;
    bool set_TS_strand_tag;

    CmdParser(int argc, char *argv[]):
        min_anchor_length(DEFAULT_MIN_ANCHOR_LENGTH),
        min_intron_length(DEFAULT_MIN_INTRON_LENGTH),
        max_intron_length(DEFAULT_MAX_INTRON_LENGTH),
        allow_anchor_indels(DEFAULT_ALLOW_ANCHOR_INDELS),
        max_anchor_indel_size(DEFAULT_MAX_ANCHOR_INDEL_SIZE),
        min_confidence_score(DEFAULT_MIN_CONFIDENCE_SCORE),
        min_read_count(DEFAULT_MIN_READ_COUNT),
        junction_filter(NULL_SJ_FILTER),
        strandness(DEFAULT_STRANDED),
        excludes(EXCLUDE_NONE),
        skip_missing_targets(false),
        set_XS_strand_tag(false),
        set_TS_strand_tag(false) {

        try {
            parse_cmd_args(argc, argv);
        } catch (const std::exception& ex) {
            cerr << "Error parsing command line: " << ex.what() << endl;
            exit(1);
        }
    }

    private:
    int parse_options(int argc, char *argv[]) {
        // definitions for long-only options
        static const int OPT_SET_XS_STRAND_TAG = 256;
        static const int OPT_SET_TS_STRAND_TAG = 257;
        
        struct option long_options[] = {
            {"help", no_argument, NULL, 'h'},
            {"version", no_argument, NULL, 'v'},
            {"unsorted", no_argument, NULL, 'u'},
            {"min-anchor-length", required_argument, NULL, 'a'},
            {"min-intron-length", required_argument, NULL, 'i'},
            {"max-intron-length", required_argument, NULL, 'I'},
            {"allow-anchor-indels", no_argument, NULL, 'd'},
            {"max-anchor-indel-size", required_argument, NULL, 'm'},
            {"min-confidence-score", required_argument, NULL, 'C'},
            {"min-read-count", required_argument, NULL, 'r'},
            {"strandness", required_argument, NULL, 's'},
            {"excludes", required_argument, NULL, 'X'},
            {"genome-fasta", required_argument, NULL, 'g'},
            {"sj-filter", required_argument, NULL, 'f'},
            {"skip-missing-targets", no_argument, NULL, 'S'},
            {"junction-bed", required_argument, NULL, 'j'},
            {"intron-bed", required_argument, NULL, 'n'},
            {"intron-bed6", required_argument, NULL, 'b'},
            {"intron-calls", required_argument, NULL, 'c'},
            {"pass-through", required_argument, NULL, 'p'},
            {"debug-trace", required_argument, NULL, 'D'},
            {"set-XS-strand-tag", no_argument, NULL, OPT_SET_XS_STRAND_TAG},
            {"set-TS-strand-tag", no_argument, NULL, OPT_SET_TS_STRAND_TAG},
            {NULL, 0, NULL, 0}
        };
            
        const char *short_options = "hvua:i:I:C:r:s:X:g:S:j:n:b:c:p:D:f:";
        int c;
        while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
            switch (c) {
                case 'a':
                    min_anchor_length = atoi(optarg);
                    break;
                case 'u':
                    unsorted = true;
                    break;
                case 'd':
                    allow_anchor_indels = true;
                    break;
                case 'm':
                    max_anchor_indel_size = toUnsigned(optarg);
                    break;
                case 'i':
                    min_intron_length = toUnsigned(optarg);
                    break;
                case 'I':
                    max_intron_length = toUnsigned(optarg);
                    break;
                case 'C':
                    min_confidence_score = toFloat(optarg);
                    break;
                case 'r':
                    min_read_count = toUnsigned(optarg);
                    break;
                case 's':
                    strandness = str_to_strandness(optarg);
                    break;
                case 'X':
                    excludes |= parse_exclude_cat(optarg);
                    break;
                case 'g':
                    genome_fa = string(optarg);
                    break;
                case 'f':
                    junction_filter = junction_filter_parse(string(optarg));
                    break;
                case 'S':
                    skip_missing_targets = true;
                    break;
                case 'j':
                    junction_bed = optarg;
                    break;
                case 'n':
                    intron_bed = optarg;
                    break;
                case 'b':
                    intron_bed6 = optarg;
                    break;
                case 'c':
                    intron_call_tsv = optarg;
                    break;
                case 'p':
                    bam_pass_through = optarg;
                    break;
                case 'D':
                    debug_trace_tsv = optarg;
                    break;
                case OPT_SET_XS_STRAND_TAG:
                    set_XS_strand_tag = true;
                    break;
                case OPT_SET_TS_STRAND_TAG:
                    set_TS_strand_tag = true;
                    break;
                case 'h':
                    usage();
                case 'v':
                    cerr << PACKAGE_NAME << " " << PACKAGE_VERSION << " " << PACKAGE_URL << endl;
                    exit(0);
                case '?':
                default:
                    cmd_error("bug: option not handled");
            }
        }
        return optind;
    }

    void collect_bams(int argc, char *argv[], int nextopt) {
        while (nextopt < argc) {
            bam_files.push_back(string(argv[nextopt++]));
        }
        if (bam_files.size() == 0) {
            bam_files.push_back("/dev/stdin");
        }
        if ((bam_files.size() > 1) and (bam_pass_through != "")) {
            cmd_error("can only have one input BAM with --bam-pass-through");
        }
    }

    void setup_junction_filter() {
        if ((junction_filter == CANON_SJ_FILTER) and (genome_fa == "")) {
            cmd_error("--sj-filter=canon requires --genome_fa");
        }
        if ((junction_filter == NULL_SJ_FILTER) and (genome_fa != "")) {
            // default to canon with genome
            junction_filter = CANON_SJ_FILTER;
        }
    }

    void parse_cmd_args(int argc, char *argv[]) {
        int nextopt = parse_options(argc, argv);
        collect_bams(argc, argv, nextopt);
        setup_junction_filter();
        
        // if indels are not allowed, also indicate this by setting max size
        if (not allow_anchor_indels) {
            max_anchor_indel_size = 0;
        }

        // can only sets strand on pass through with genome
        if ((set_XS_strand_tag or set_TS_strand_tag) and
            ((bam_pass_through.size() == 0) or (genome_fa.size() == 0))) {
            cmd_error("--set-XS-strand-tag and --set-TS-strand-tag require --pass-through and --genome-fasta");
        }
    }
};


class OutputFiles {
    public:
    const string bam_pass_through;
    ostream* junction_bed_fh;
    ostream* intron_bed_fh;
    ostream* intron_bed6_fh;
    ostream* intron_call_fh;

    OutputFiles(const CmdParser& opts):
        bam_pass_through(opts.bam_pass_through),
        junction_bed_fh(open_out_or_null(opts.junction_bed)),
        intron_bed_fh(open_out_or_null(opts.intron_bed)),
        intron_bed6_fh(open_out_or_null(opts.intron_bed6)),
        intron_call_fh(open_out_or_null(opts.intron_call_tsv)) {
        if (intron_call_fh != NULL) {
            print_junction_call_header(*intron_call_fh);
        }
    }

    ~OutputFiles() {
        delete junction_bed_fh;
        delete intron_bed_fh;
        delete intron_bed6_fh;
        delete intron_call_fh;
    }
};

static void output_junctions(JunctionsExtractor& extractor,
                             float min_confidence_score,
                             uint32_t min_read_count,
                             OutputFiles& output) {
    JunctionVector juncs = extractor.get_junctions();
    if (output.junction_bed_fh != NULL) {
        juncs.sort_by_anchors();
        print_anchor_bed(juncs, min_confidence_score, min_read_count, *output.junction_bed_fh);
    }
    if ((output.intron_bed_fh != NULL) or (output.intron_bed6_fh != NULL)
        or (output.intron_call_fh != NULL)) {
        juncs.sort_by_introns();
        if (output.intron_bed_fh != NULL) {
            print_intron_bed(juncs, min_confidence_score, min_read_count, 9, *output.intron_bed_fh);
        }
        if (output.intron_bed6_fh != NULL) {
            print_intron_bed(juncs, min_confidence_score, min_read_count, 6, *output.intron_bed6_fh);
        }
        if (output.intron_call_fh != NULL) {
            print_intron_call_tsv(juncs, min_confidence_score, min_read_count, *output.intron_call_fh);
        }
    }
}

/* Extract junctions for the single BAM, optimizing memory,
 * by process one target at a time.
 */
static void extract_junctions_by_target(JunctionsExtractor& extractor,
                                        const string& bam_file,
                                        float min_confidence_score,
                                        uint32_t min_read_count,
                                        OutputFiles& output) {
    extractor.open(bam_file);
    if (output.bam_pass_through != "") {
        extractor.open_pass_through(output.bam_pass_through);
    }
    for (int target_index = 0; target_index < extractor.get_num_targets(); target_index++) {
        extractor.identify_junctions_for_target(target_index);
        output_junctions(extractor, min_confidence_score, min_read_count, output);
        extractor.clear();
    }
    extractor.copy_unaligned_reads();
    extractor.close();
}

/* Extract junctions for an unsorted BAM.  This allows pass-through.
 */
static void extract_junctions_unsorted(JunctionsExtractor& extractor,
                                       const string& bam_file,
                                       float min_confidence_score,
                                       uint32_t min_read_count,
                                       OutputFiles& output) {
    extractor.open(bam_file);
    if (output.bam_pass_through != "") {
        extractor.open_pass_through(output.bam_pass_through);
    }
    extractor.identify_junctions_for_bam();
    extractor.copy_unaligned_reads();
    extractor.close();
    output_junctions(extractor, min_confidence_score, min_read_count, output);
}
                                
/* Extract junctions for the multiple BAM files.
 */
static void extract_junctions_multiple(JunctionsExtractor& extractor,
                                       const vector<string>& bam_files,
                                       float min_confidence_score,
                                       uint32_t min_read_count,
                                       OutputFiles& output) {
    for (int ibam = 0; ibam < bam_files.size(); ibam++) {
        extractor.open(bam_files[ibam]);
        extractor.identify_junctions_for_bam();
        extractor.close();
    }
    output_junctions(extractor, min_confidence_score, min_read_count, output);
}
                                
static void extract_junctions(CmdParser &opts) {
    Genome *genome = NULL;
    if (opts.genome_fa.size() > 0) {
        genome = new Genome(opts.genome_fa);
    }
    ostream *trace_fh = open_out_or_null(opts.debug_trace_tsv);
    JunctionsExtractor extractor(opts.min_anchor_length, opts.max_anchor_indel_size,
                                 opts.min_intron_length, opts.max_intron_length,
                                 opts.junction_filter,
                                 opts.strandness, opts.excludes, genome,
                                 opts.skip_missing_targets,
                                 opts.set_XS_strand_tag, opts.set_TS_strand_tag,
                                 trace_fh);
    OutputFiles output(opts);
    if (opts.bam_files.size() > 1) {
        extract_junctions_multiple(extractor, opts.bam_files,
                                   opts.min_confidence_score, opts.min_read_count, output);
    } else if (opts.unsorted) {
        extract_junctions_unsorted(extractor, opts.bam_files[0],
                                   opts.min_confidence_score, opts.min_read_count, output);
    } else {
        extract_junctions_by_target(extractor, opts.bam_files[0],
                                    opts.min_confidence_score, opts.min_read_count, output);
    }
    delete trace_fh;
    delete genome;
 }

// entry point
int main(int argc, char *argv[]) {
    CmdParser opts(argc, argv);
    try {
        extract_junctions(opts);
    } catch (const std::exception& e) {
        cmd_error(e.what());
    }
    return 0;
}
